File: Data Structures & Algorithms/trapping-rain-water/submission-3.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] rightmax = new int[n];
        for(int i=n-2; i>=0; i--) {
            rightmax[i] = Math.max(height[i+1], rightmax[i+1]);
        }
        int leftmax = 0;
        int i=0;
        while(i<n && height[i]>=leftmax) {
            leftmax = height[i];
            i++;
        }
        int ans = 0;
        while(i<n) {
            if(leftmax>height[i] && rightmax[i]>height[i]) {
                int ht = Math.min(leftmax, rightmax[i]);
                ans += ht-height[i];
            }
            leftmax = Math.max(leftmax, height[i]);
            i++;
        }
        return ans;
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-3.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = []
        if len(intervals)==0:
            return [newInterval]
        for i in range(len(intervals)):
            intr = intervals[i]
            if newInterval[1]<intr[0]:
                ans.append(newInterval)
                return ans + intervals[i:]
            if newInterval[0]<=intr[1]:
                newInterval[0] = min(newInterval[0],intr[0])
                newInterval[1] = max(newInterval[1],intr[1])
            else:
                ans.append(intr)
        ans.append(newInterval)
        return ans